### PR TITLE
Bundling 2.2.4

### PR DESCRIPTION
**Pull request by kandy**

### CODE REVIEW

File: Block/Adminhtml/Dashboard/Information.php [hunk 1]
```diff
@@ -143,8 +143,6 @@ public function getDynamicContentPasscode()
      */
     public function getAbandonedCartLimit()
     {
-        return ($this->data->getAbandonedCartLimit())?
-            '<span class="message message-warning">' . $this->escapeHtml($this->data->getAbandonedCartLimit()).
-            ' h</span>' : 'No limit';
+        return ($this->data->getAbandonedCartLimit())? $this->data->getAbandonedCartLimit() . ' h': __('No limit');
```
Comment (kandy) [inline Block/Adminhtml/Dashboard/Information.php]: :warning: Please add translate for getAbandonedCartLimit() like __('%1 h',$this->data->getAbandonedCartLimit() )

File: Controller/Adminhtml/Abandoned.php [hunk 2]
```diff
@@ -0,0 +1,14 @@
+<?php
+
+namespace Dotdigitalgroup\Email\Controller\Adminhtml;
+
+abstract class Abandoned extends \Magento\Backend\App\Action
```
Comment (kandy) [inline Controller/Adminhtml/Abandoned.php:5]: ⚠️ Inheritance is not recommended. Please define ADMIN_RESOURCE constant in every child class

File: Model/Connector/Order.php [hunk 3]
```diff
@@ -638,10 +637,10 @@ public function _getAttributesArray($attributeSetId)
      *
      * @return string
      */
-    public function _limitLength($value)
+    private function limitLength($value)
     {
-        if (strlen($value) > 250) {
-            $value = substr($value, 0, 250);
+        if (mb_strlen($value) > 250) {
```
Comment (kandy) [inline Model/Connector/Order.php]: Please use \Magento\Framework\Stdlib\StringUtils here.
Also, use const for 250 magic number

File: Model/Connector/Product.php [hunk 4]
```diff
@@ -175,8 +175,8 @@ public function setProduct($product)
 
         $shortDescription = $product->getShortDescription();
         //limit short description
-        if (strlen($shortDescription) > 250) {
-            $shortDescription = substr($shortDescription, 0, 250);
+        if (mb_strlen($shortDescription) > 250) {
```
Comment (kandy) [inline Model/Connector/Product.php]: Please use \Magento\Framework\Stdlib\StringUtils here

File: Model/Connector/Product.php [hunk 5]
```diff
@@ -338,4 +336,18 @@ public function __sleep()
     public function __wakeup()
     {
     }
+
+    /**
+     * @param $productAttribute array
```
Comment (kandy) [inline Model/Connector/Product.php]: Add method description

File: Model/Connector/Product.php [hunk 6]
```diff
@@ -338,4 +336,18 @@ public function __sleep()
...
+     * @return string
+     */
+    private function validateProductLabel($productAttribute)
```
Comment (kandy) [inline Model/Connector/Product.php]: This method transform label, not validate. Please rename

File: Model/Cron.php [hunk 7]
```diff
@@ -305,4 +361,17 @@ public function cleaning()
 
         return $message;
     }
+
+    /**
+     * @param $jobCode
+     * @return bool
+     */
+    private function jobHasAlreadyBeenRun($jobCode)
```
Comment (kandy) [inline Model/Cron.php]: Please add description to method and describe why it needs

File: Model/Importer.php [hunk 8]
```diff
@@ -220,26 +220,34 @@ public function registerQueue(
...
+                $this->importerResource->save($this);
 
-            return true;
+                return true;
+            }
         } catch (\Exception $e) {
             $this->helper->debug((string)$e, []);
         }
 
+        if (json_last_error_msg() != "No error") {
```
Comment (kandy) [inline Model/Importer.php]: ⚠️ Move this code to place where json_decode is called

File: Model/Newsletter/Subscriber.php [hunk 9]
```diff
@@ -240,8 +248,10 @@ public function unsubscribe()
             // Contacts to un-subscribe
             foreach ($contacts as $apiContact) {
                 if (isset($apiContact->suppressedContact)) {
-                    $suppressedContact = $apiContact->suppressedContact;
-                    $suppressedEmails[] = $suppressedContact->email;
+                    $suppressedContactEmail = $apiContact->suppressedContact->email;
+                    if (!in_array($suppressedContactEmail, $suppressedEmails)) {
```
Comment (kandy) [inline Model/Newsletter/Subscriber.php]: Better to use $strict comparison if posible

File: Model/ResourceModel/Automation/Collection.php [hunk 10]
```diff
@@ -66,4 +66,22 @@ public function getCollectionByType($type, $limit)
 
         return $collection;
     }
+
+    /**
+     * @param $typeId
+     * @param $automationTypeId
+     * @return bool
+     */
+    public function hasAutomationEnrolmentExists($typeId, $automationTypeId)
```
Comment (kandy) [inline Model/ResourceModel/Automation/Collection.php]: :exclamation: This method cannot  be called more than one times

File: Model/ResourceModel/Contact.php [hunk 11]
```diff
@@ -819,58 +791,61 @@ private function buildColumnData(
...
-        return $mostData;
+        return $mostData = new \Zend_Db_Expr(
+            "(
+                SELECT eaov.option_id from $salesOrder sfo
+                LEFT JOIN $salesOrderItem as sfoi on sfoi.order_id = sfo.entity_id
+                LEFT JOIN $catalogProductEntityInt pei on $leftJoinOnSfoiProductId = sfoi.product_id
+                LEFT JOIN $eavAttribute ea ON pei.attribute_id = ea.attribute_id
+                LEFT JOIN $eavAttributeOptionValue as eaov on pei.value = eaov.option_id
+                $where
+                AND ea.attribute_code = '$this->brand'
```
Comment (kandy) [inline Model/ResourceModel/Contact.php]: ⚠️  brand variable  should be properly quoted

File: Model/ResourceModel/Contact.php [hunk 12]
```diff
@@ -819,58 +791,61 @@ private function buildColumnData(
...
+        );
+    }
+
+    /**
+     * Set brand attribute
+     *
+     * @param $brand
+     * @return $this
+     */
+    public function setBrandAttribute($brand)
```
Comment (kandy) [inline Model/ResourceModel/Contact.php]: ⚠️ You should avoid state and temporary coupling in service class

File: Model/ResourceModel/Contact.php [hunk 13]
```diff
@@ -1034,4 +1009,18 @@ public function getDateLastCronRun($cronJob)
 
         return $executedAt;
     }
+
+    /**
+     * Set contacts to re-import by customer ids
+     *
+     * @param $customerIds
+     */
+    public function setNotImportedByCustomerIds($customerIds)
```
Comment (kandy) [inline Model/ResourceModel/Contact.php]: Please update method name to 'update...'

File: Model/ResourceModel/Cron/Collection.php [hunk 14]
```diff
@@ -19,4 +19,35 @@ public function _construct()
     {
         $this->_init(\Magento\Cron\Model\Schedule::class, \Magento\Cron\Model\ResourceModel\Schedule::class);
     }
+
+    /**
+     * @param $jobCode
+     * @return \Magento\Cron\Model\ResourceModel\Schedule\
+     */
+    public function getRunningJobByCode($jobCode)
```
Comment (kandy) [inline Model/ResourceModel/Cron/Collection.php]: :exclamation: This method cannot  be called more than one times

File: Model/ResourceModel/Cron/Collection.php [hunk 15]
```diff
@@ -19,4 +19,35 @@ public function _construct()
...
+            ->getFirstItem();
+    }
+
+    /**
+     * @param $jobCode
+     * @param $scheduledAt
+     *
+     * @return bool
+     */
+    public function jobOfSameTypeAndScheduledAtDateAlreadyExecuted($jobCode, $scheduledAt)
```
Comment (kandy) [inline Model/ResourceModel/Cron/Collection.php]: :exclamation: This method cannot  be called more than one times

File: Model/ResourceModel/Rules.php [hunk 16]
```diff
@@ -16,6 +22,32 @@ public function _construct()
...
 
+    /**
+     * Rules constructor.
+     *
+     * @param Json $serializer
+     * @param \Magento\Framework\Model\ResourceModel\Db\Context $context
+     * @param null $connectionName
+     */
+    public function __construct(
+        Json $serializer,
```
Comment (kandy) [inline Model/ResourceModel/Rules.php:33]: Better to preserve the order or argument

File: Model/ResourceModel/Contact.php [hunk 17]
```diff
@@ -819,58 +791,61 @@ private function buildColumnData(
...
         $salesOrder,
         $salesOrderItem,
         $catalogProductEntityInt,
         $eavAttribute,
-        $eavAttributeOptionValue
+        $eavAttributeOptionValue,
+        $forSubscriber = false
     ) {
+        if (! $this->brand) {
+            return new \Zend_Db_Expr('NULL');
```
Comment (kandy) [inline Model/ResourceModel/Contact.php]: Use \Magento\Framework\DB\Sql\Expression here

File: Model/ResourceModel/Contact.php [hunk 18]
```diff
@@ -819,58 +791,61 @@ private function buildColumnData(
...
-                    AND eaov.value is not null
-                    GROUP BY eaov.value
-                    HAVING count(*) > 0
-                    ORDER BY count(*) DESC
-                    LIMIT 1
-                )"
-            );
-        }
-        return $mostData;
+        return $mostData = new \Zend_Db_Expr(
```
Comment (kandy) [inline Model/ResourceModel/Contact.php]: Use \Magento\Framework\DB\Sql\Expression here

File: Block/Adminhtml/Config/Developer/EmailTemplatesSync.php [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+
+    /**
+     * Get the button and scripts contents.
+     *
+     * @param \Magento\Framework\Data\Form\Element\AbstractElement $element
+     *
+     * @return string
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function _getElementHtml(\Magento\Framework\Data\Form\Element\AbstractElement $element)
```
Comment (kandy) [inline Block/Adminhtml/Config/Developer/EmailTemplatesSync.php]: Mark method  as protected

File: Block/Adminhtml/Config/Developer/Templatesync.php [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+
+    /**
+     * Get the button and scripts contents.
+     *
+     * @param \Magento\Framework\Data\Form\Element\AbstractElement $element
+     *
+     * @return string
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function _getElementHtml(\Magento\Framework\Data\Form\Element\AbstractElement $element)
```
Comment (kandy) [inline Block/Adminhtml/Config/Developer/Templatesync.php]: Mark method protected

File: Model/ResourceModel/Automation/Collection.php (hunk 10 above)
Comment (adeelq) [reply Model/ResourceModel/Automation/Collection.php]: @kandy Why is this a problem?

File: Model/ResourceModel/Cron/Collection.php (hunk 14 above)
Comment (adeelq) [reply Model/ResourceModel/Cron/Collection.php]: Same here @kandy Why is this a problem?

File: Model/ResourceModel/Cron/Collection.php (hunk 15 above)
Comment (adeelq) [reply Model/ResourceModel/Cron/Collection.php]: Same here @kandy Why is this a problem?

File: Model/ResourceModel/Rules.php (hunk 16 above)
Comment (adeelq) [reply Model/ResourceModel/Rules.php:33]: @kandy Not sure if i understand correctly. Can you please explain more?

Comment (simon-letch) [reply Model/ResourceModel/Rules.php:33]: The base class constructor takes the args $context, $connectionName in that order. These should be the first two parameters in this class i think so $serializer comes last.

Comment (simon-letch) [reply Model/ResourceModel/Rules.php:33]: not sure it's possible though with the default argument value for $connectionName?

Comment (adeelq) [reply Model/ResourceModel/Rules.php:33]: indeed thats why $connectionName is the last param since it has default value

Comment (adeelq) [reply Model/ResourceModel/Rules.php:33]: @kandy Never mind. I understand what you mean

File: Model/ResourceModel/Automation/Collection.php (hunk 10 above)
Comment (kandy) [reply Model/ResourceModel/Automation/Collection.php]: The problem that you should know the implementation of a method to use it. 
Also this violates the 2.12. item of [technical guidelines](http://devdocs.magento.com/guides/v2.2/coding-standards/technical-guidelines.html).

File: Model/ResourceModel/Cron/Collection.php (hunk 14 above)
Comment (kandy) [reply Model/ResourceModel/Cron/Collection.php]: see https://github.com/kandy/dotmailer-magento2-extension/pull/1#discussion_r169621556

File: Model/ResourceModel/Cron/Collection.php (hunk 15 above)
Comment (kandy) [reply Model/ResourceModel/Cron/Collection.php]: see https://github.com/kandy/dotmailer-magento2-extension/pull/1#discussion_r169621556

File: Model/ResourceModel/Automation/Collection.php (hunk 10 above)
Comment (adeelq) [reply Model/ResourceModel/Automation/Collection.php]: @kandy We do think you point is right however not for above function. Maybe you can provide us one example to clarify the issue please.

Comment (kandy) [reply Model/ResourceModel/Automation/Collection.php]: Example pseudocode:
``` 
$collection =   $this->objectManger->create(\Dotdigitalgroup\Email\Model\ResourceModel\Automation\Collection::class);

if ($collection->hasAutomationEnrolmentExists($typeId, $automationTypeId)) {
     $collection->load(); // <-- will load only one element, but it unexpected result, I did not call any setters  
}
```
To resolve this issue I recommend to extract this type of functions in separate service class

Comment (adeelq) [reply Model/ResourceModel/Automation/Collection.php]: Thanks

**Pull request closed without merging**